docs/sd-settings.md
# Stokesian Dynamics settings documentation

Before reading this documentation, make sure you have read the repository [README](../README.md).

## Introduction
The **settings.py** file contains various configuration settings for the Stokesian Dynamics simulation software in the **stokesian_dynamics** folder.

## Basic settings

### `setup_number`
- Type: `int`
- Default value: `1`
- Description: References which initial particle position setup to use, which includes the number of particles, the size of the particles, and the the type of the particle (sphere or dumbbell). The setup number corresponds to a specific particle configuration in **setups/positions.py**. This file can be edited to add new setups.

This can be overridden from the command line by passing the setup number as the first argument.

### `input_number`
- Type: `int`
- Default value: `1`
- Description: References which forces and/or velocities to impose on the particles (the right-hand side of the mobility problem). The input number corresponds to a specific force and velocity configuration in **setups/inputs.py**. This file can be edited to add new force/velocity setups.

This can be overridden from the command line by passing the input number as the second argument.

### `fully_2d_problem`
- Type: `bool`
- Default value: `False`
- Description: Specifies whether the simulation is a fully 2D problem. If set to `True`, the simulation will always set the value of the particles' *y*-coordinates to 0.

### `timestep`
- Type: `float`
- Default value: `0.1`
- Description: Specifies the timestep size for the simulation. Smaller values result in more accurate simulations but can increase computational cost if you need more of them.

This can be overridden from the command line by passing the timestep as the third argument.

### `num_frames`
- Type: `int`
- Default value: `100`
- Description: Specifies the number of frames to run the simulation for. This value determines the duration of the simulation.

This can be overridden from the command line by passing the number of frames as the fourth argument.

### `input_form`
- Type: `str`
- Default value: `'fte'`
- Description: Specifies the type of force and/or velocity to impose on the particles, i.e. what goes on the left and right-hand sides of the mobility problem. The available options are:
  - `'fte'`: 
    - Input: Forces, torques and background rate of strain
    - Output: Velocities, angular velocities and stresslets
  - `'fts'`: 
    - Input: Forces, torques and stresslets
    - Output: Velocities, angular velocities and background rate of strain
  - `'ufte'`: 
    - Input: Velocities of some spheres; forces on the rest of the spheres; torques and background rate of strain for all spheres; forces on the dumbbells
    - Output: Forces on the fixed-velocity spheres; velocities of the free spheres; angular velocities and stresslets for all spheres; velocities of the dumbbells
  - `'ufteu'`: 
    - Input: Velocities of some spheres; forces on the rest of the spheres; torques and background rate of strain for all spheres; velocities of some dumbbells; forces on the rest of the dumbbells
    - Output: Forces on the fixed-velocity spheres; velocities of the free spheres; angular velocities and stresslets for all spheres; forces on the fixed-velocity dumbbells; velocities for the free dumbbells
  - `'duf'`: 
    - Input: (Dumbbells only) Velocities of some dumbbells; forces on the rest of the dumbbells
    - Output: Forces on the fixed-velocity dumbbells; velocities of the free dumbbells

This can be overridden from the command line by passing the input form as the fifth argument.

## Advanced settings

### `invert_m_every`
- Type: `int`
- Default value: `1`
- Description: Specifies how many timesteps to wait in between finding the inverse of Minfinity, the far-field mobility matrix. In between these timesteps, the previously computed value is used. Computing Minfinity can be the most computationally expensive part of the simulation, so setting this value to a higher number can reduce the computational cost.

### `cutoff_factor`
- Type: `float`
- Default value: `2`
- Description: Specifies when to start using R2bexact. The cutoff factor is the ratio of the distance between the particle centres to the sum of the particle radii.

### `timestepping_scheme`
- Type: `str`
- Default value: `'euler'`
- Description: Specifies the timestepping scheme to use. The available options are:
  - `'euler'`: Euler method
  - `'ab2'`: Adams–Bashforth 2nd order method
  - `'rk4'`: Runge–Kutta 4th order method

### `rk4_generate_minfinity_for_each_stage`
- Type: `bool`
- Default value: `True`
- Description: Specifies whether to work out Minfinity for each of the 4 stages when using the Runge–Kutta 4th order method, or to approximate them by just using the Minfinity from the first stage. In the absence of R2Bexact, setting this to `False` makes the timestep equivalent to the Euler method. If R2Bexact is present and `invert_m_every` is greater than 1, choosing `False` might be an OK approximation.

### `feed_every_n_timesteps`
- Type: `int`
- Default value: `0` (off)
- Description: Specifies how often to feed in new particles to a simulation. Useful if the simulation involves large domains where an object of interest moves through a sea of obstacles, and you only need to actually include the obstacles when the object gets near. This will require some configuration in `run_simulation.py`: just search for `feed_every_n_timesteps` in the source code. If set to `0`, feeding is turned off.

### `feed_from_file`
- Type: `str`
- Default value: None
- Description: Specifies the name of the file to feed new particles from. This is only relevant if `feed_every_n_timesteps` is greater than 0.

### `how_far_to_reproduce_gridpoints`
- Type: `int`
- Default value: `2`
- Description: Specifies how many repeats of the box to consider before the contribution decays away for periodic boxes. If the box width is `Lx`, `1` corresponds to `[-Lx, 0, Lx]`, `2` corresponds to `[-2Lx, -Lx, 0, Lx, 2Lx]`, and so on. `2` is normally sufficient, but `3` gives a bit more accuracy.

### `printout`
- Type: `int`
- Default value: `0`
- Description: Specifies the level of output to the screen. The available options are:
  - `0`: Minimal output
  - `1`: Separation distance and velocities
  - `2`: Matrices
  - `3`: Individual calculations

### `email_on_completion`
- Type: `bool`
- Default value: `False`
- Description: Specifies whether to send an email on completion of the simulation. This requires setting up the email settings in **functions/email.py**.

### `view_graphics`
- Type: `bool`
- Default value: `True`
- Description: Specifies whether to view graphics during the simulation. If set to `True`, the simulation will display the particles' positions and velocities in a 3D plot.

### `bead_bead_interactions`
- Type: `bool`
- Default value: `True`
- Description: Specifies whether to include interactions between dumbbell beads in the simulation. This should really always be set to `True`.

### `explosion_protection`
- Type: `bool`
- Default value: `False`
- Description: Specifies whether to include 'explosion protection' in the simulation. Specifically, dumbbells with a separation distance greater than 5 times the default sphere radius (i.e. 5 in simulation units) will stop the execution of the simulation. Typically this would suggest something has gone wrong.

### `save_positions_every_n_timesteps`
- Type: `int`
- Default value: `1`
- Description: Specifies how often to save the particles' positions to a file. If set to `1`, the positions will be saved every timestep.

### `save_forces_every_n_timesteps`
- Type: `int`
- Default value: `1`
- Description: Specifies how often to save the forces on the particles to a file. If set to `1`, the forces will be saved every timestep.

### `save_forces_and_positions_to_temp_file_as_well`
- Type: `bool`
- Default value: `True`
- Description: Specifies whether to save the forces and positions to a temporary file as well as to the main output file. This is useful if the simulation is expected to run for a long time and you want to make a running backup in case the power fails, allowing you to restart the simulation from the last saved state.

### `save_to_temp_file_every_n_timesteps`
- Type: `int`
- Default value: `120`
- Description: Specifies how often to save the forces and positions to a temporary file. If set to `120`, the forces and positions will be saved at least every 120 timesteps, if `save_positions_every_n_timesteps` is not smaller.

### `extract_force_on_wall_due_to_dumbbells`
- Type: `bool`
- Default value: `False`
- Description: Specifies whether to extract the total force on any fixed particles (typically useful when fixed particles are representing a wall) due to the dumbbells. If set to `True`, the force on the wall due to the dumbbells will be saved to the output NPZ file as `force_on_wall_due_to_dumbbells`.

### `use_drag_Minfinity`
- Type: `bool`
- Default value: `False`
- Description: Specifies whether to use the drag Minfinity, i.e. diagonal self-terms only. This might a good approximation for dense suspensions only.

### `use_Minfinity_only`
- Type: `bool`
- Default value: `False`
- Description: Specifies whether to use Minfinity only, i.e. turn off R2Bexact. This is mostly put in for a little discussion.

### `use_drag_R2Binfinity`
- Type: `bool`
- Default value: `False`
- Description: Only relevant if `use_drag_Minfinity = True`. The precomputed XYZ scalars in the resistance matrix already have R2Binfinity subtracted off them (it's quicker). These data files are labelled with '-d'. If `use_drag_Minfinity` is set to `True`, there is a choice of using '-d' scalars where the drag-only R2Binfinity is subtracted, or using '-d' scalars where the full R2Binfinity is subtracted. You might pick the drag version if you wanted to make sure that the two-sphere case is consistent, but you might pick the full version if you have a dense suspension.

### `config.DISABLE_JIT`
- Type: `bool`
- Default value: `True` (i.e. Numba is disabled)
- Description: Specifies whether to turn Numba on (`False`) or off (`True`).

## Graphical settings

These are only relevant if `view_graphics` is set to `True`.

### `viewbox_bottomleft_topright`
- Type: `np.ndarray`
- Default value: `np.array([[-15, 0, -15], [15, 1, 15]])`
- Description: Specifies the bottom-left and top-right corners of the viewbox for the 3D plot. The viewbox is the region of space that is visible in the 3D plot.

### `viewing_angle`
- Type: `tuple`
- Default value: `(0, -90)`
- Description: Specifies the viewing angle for the 3D plot. The first value is the elevation angle and the second value is the azimuthal angle. `(0,-90)` represents the x-z plane; `(30,-60)` is the matplotlib default.

### `view_labels`
- Type: `bool`
- Default value: `False`
- Description: Specifies whether to view arrows on the spheres in the 3D plot representing forces. If set to `True`, arrows will be displayed.

### `trace_paths`
- Type: `int`
- Default value: `0` (off)
- Description: Specifies whether to trace the paths of the particles in the 3D plot. If set to `n` > 0, a line will be drawn between every `n` timesteps.

### `two_d_plot`
- Type: `bool`
- Default value: `True`
- Description: Specifies whether to remove the third axis in the plot and display as a 2D plot instead.




stokesian_dynamics/settings.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Adam Townsend, adam@adamtownsend.com, 07/06/2017

"""Settings for a Stokesian Dynamics simulation.

See docs/settings.md for a full explanation of each setting.

This file is imported by all other scripts in the Stokesian Dynamics package,
so it is the place to set up your simulation. You can override the inputs
    setup_number, input_number, timestep, num_frames
by passing them in as arguments to run_simulation.py from the command line.
See the README for more information.
"""

import sys
import numpy as np
from numba import config
from functions.shared import throw_error


# ----------------------------------------------------------------------------|
# BASIC SETTINGS

setup_number = 1
input_number = 1
fully_2d_problem = False

# Timestep (default = 0.1)
timestep = 0.1

# Number of frames to run simulation for
num_frames = 100

# How to solve the equation U=MF. {'fts','fte','ufte','ufteu','duf'}
input_form = 'fte'

# ----------------------------------------------------------------------------|
# BASIC SETTINGS: OVERRIDES FROM COMMAND LINE

args = sys.argv[1:]
number_of_args = len(args)
running_on_legion = 0

try:
    if number_of_args >= 6:
        extract_force_on_wall_due_to_dumbbells = (
            args[5] in ["True", "true", "t", "T", "1"])
    if number_of_args >= 5:
        input_form = args[4]
    if number_of_args >= 4:
        num_frames = int(args[3])
    if number_of_args >= 3:
        timestep = float(args[2])
    if number_of_args >= 2:
        input_number = int(args[1])
    if number_of_args >= 1:
        setup_number = int(args[0])
except ValueError:
    throw_error("Invalid input. Command-line input should come in the "
                + "order: setup number, input number, timestep, no. frames, "
                + "input form, extract dumbbell force Boolean.")

# ----------------------------------------------------------------------------|
# ADVANCED SETTINGS

# How many timesteps to wait in between finding (Minfinity)^-1 (10 is normally
# OK). In between these timesteps we use the previously computed value.
invert_m_every = 1

# When to start using R2bexact. cutoff_factor = r*/(a1+a2). Default 2.
cutoff_factor = 2

# Timestepping scheme: Choose explicit timestep from: ['euler', 'ab2', 'rk4']
# (ab2=Adams Bashforth)
timestepping_scheme = 'euler'

# If using RK4, do you want to generate Minfinity for each of the 4 stages
# (True), or just for the initial stage (False)? In the absence of R2Bexact,
# choosing False makes the timestep equivalent to Euler; but if
# R2Bexact is present, and invert_m_every>1, this might be an OK approximation.
rk4_generate_minfinity_for_each_stage = True

# Are we going to be feeding in new particles underneath the box? (Really just
# for certain types of simulations)
feed_every_n_timesteps = 0  # 0 to turn off
feed_from_file = 'feed-from-file-name'

# For periodic boxes, how many repeats of the box do we want to consider before
# the contribution decays away?
# If Lx is box width, '1' corresponds to [-Lx, 0, Lx]. '2' corresponds to
# [-2Lx, -Lx, 0, Lx, 2Lx] etc.
# '2' is normally sufficient. '3' gives you a bit more accuracy.
how_far_to_reproduce_gridpoints = 2

# Set level of output to screen. # 0 = minimal output, 1 = separation distance
# and velocities, 2 = matrices, 3 = individual calculations
printout = 0

# Send email on completion? Set this up in functions/email.py.
email_on_completion = False

# View graphics? (Note: doesn't save a video. To do that, use
# plot_particle_positions_video.py after the fact.)
view_graphics = True

# Bead-bead interactions? (Should really always be true)
bead_bead_interactions = True

# Explosion protection? (Dumbbells with dx > 5 stops the execution)
explosion_protection = False

# Save positions/forces every n timesteps?
start_saving_after_first_n_timesteps = 0
save_positions_every_n_timesteps = 1
save_forces_every_n_timesteps = 1
save_forces_and_positions_to_temp_file_as_well = True
save_to_temp_file_every_n_timesteps = 120

if start_saving_after_first_n_timesteps > num_frames:
    print("WARNING: start_saving_after_first_n_timesteps > num_frames. "
          + "Saving will fail.")

# Work out the force on any particles whose velocities are specified in the
# input file. (Typical use case: these 'fixed' particles are acting as walls.)
# This is saved to `force_on_wall_due_to_dumbbells` in the output NPZ file.
extract_force_on_wall_due_to_dumbbells = False

# Just use diagonal terms of Minfinity. Good(?) approximation for dense
# suspensions only.
use_drag_Minfinity = False
# Only use Minfinity, i.e. turn off R2Bexact. Mostly put in for a little
# discussion.
use_Minfinity_only = False

# Only relevant if  use_drag_Minfinity = True.
# The precomputed XYZ scalars in the resistance matrix already have R2binfinity
# subtracted off them (it's quicker). These data files are labelled with '-d'.
# If use_drag_Minfinity = True, there is a choice of using '-d' scalars where
# the drag-only R2Binfinity is subtracted, or using '-d' scalars where the full
# R2Binfinity is subtracted. You might pick the drag version if you wanted
# to make sure that the two-sphere case works, but you might pick the full
# version if you have a dense suspension.
use_drag_R2Binfinity = False

# Turn Numba on or off
config.DISABLE_JIT = True

# ----------------------------------------------------------------------------|
# GRAPHICAL SETTINGS

# If generating video, what should we be able to see?
viewbox_bottomleft_topright = np.array([[-15, 0, -15], [15, 1, 15]])

# Viewing angle on video, (elev,azim).
# e.g. (0,-90) = x-z plane; (30,-60) = Matplotlib default.
viewing_angle = (0, -90)

# View arrows on spheres in the video?
view_labels = False

# Trace the paths on the video? 0 off, n>=1 on. It will generate a line between
# every n timesteps.
trace_paths = 0

# 2D Plot? This removes the third axis.
two_d_plot = True
